[mhlo] Fix warnings emited by `-Wsign-compare`

ashay opened the pull request:
Building with clang (v14) produces numerous warnings about comparison
between signed and unsigned values.  This patch introduces explicit cast
operations to eliminate the warnings.  The unit tests run as part of the
`check-mlir-hlo` targets pass.

gbaned commented:
@ashay Can you please sign CLA. Thank you!

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/gml_st/IR/gml_st_ops.cc:
```diff
@@ -1734,7 +1734,7 @@ LogicalResult TransposeDimsOp::verify() {
   SmallVector<int64_t> position(rank, -1);
   for (const auto &it : llvm::enumerate(permutation())) {
-    int64_t dim = it.value();
+    size_t dim = it.value();
```
I rather keep signed integer semantics whenever possible, through casting as needed: all int64_t instead of size_t are really intended.

ashay replied:
Makes sense, thanks!  I updated the patch to prefer signed integers over size_t.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/gml_st/IR/gml_st_ops.cc:
```diff
@@ -1823,14 +1824,15 @@ void SetYieldOp::build(
     (*builderFnIt)(builder, result.location, bodyBlock.getArgument(0),
                    bodyBlock.getArgument(1));
-    std::next(builderFnIt);
+    (void)std::next(builderFnIt);
```
I'm confused, isn't this hiding a real bug here? (@pifon2a )

joker-eph replied:
See: https://godbolt.org/z/Kx3aezonj

pifon2a replied:
@joker-eph Yes, it is a real bug. It was not visible because we have never built a terminator with multiple accumulators programmatically. Thank you! I sent the fix.

joker-eph replied:
@ashay can you remove all these (void) for now?

ashay replied:
Sounds good, removed.

ashay committed:
[mhlo] Fix warnings emited by -Wsign-compare

Building with clang (v14) produces numerous warnings about comparison
between signed and unsigned values. This patch introduces explicit cast
operations to eliminate the warnings. The unit tests run as part of the
`check-mlir-hlo` target pass.

gbaned commented:
@ashay Can you please resolve conflicts? Thank you!

ashay commented:
Sorry about the delay, @gbaned!  Resolved the conflict in the most recent version.

gbaned commented:
@ashay Can you please resolve conflicts? Thank you!
